File: android/src/main/java/me/orbitium/flutter_overlay_window_engine/entities/OverlayNotification.java
package me.orbitium.flutter_overlay_window_engine.entities;

import java.util.Map;

import me.orbitium.flutter_overlay_window_engine.overlay.NotificationVisibility;

public class OverlayNotification {
    private final String title;
    private final String content;
    private final NotificationVisibility notificationVisibility;

    public OverlayNotification(String title, String content, NotificationVisibility notificationVisibility) {
        this.title = title;
        this.content = content;
        this.notificationVisibility = notificationVisibility;
    }

    public static final OverlayNotification noNotification = new OverlayNotification("", "", NotificationVisibility.none);

    // Getters
    public String getTitle() {
        return title;
    }

    public String getContent() {
        return content;
    }

    public NotificationVisibility getNotificationVisibility() {
        return notificationVisibility;
    }

    public static OverlayNotification fromJson(Map<String, Object> json) {
        String title = (String) json.get("title");
        String content = (String) json.get("content");
        String visibilityIndex = (String) json.get("notificationVisibility");
        NotificationVisibility notificationVisibility = NotificationVisibility.valueOf(visibilityIndex);

        return new OverlayNotification(title, content, notificationVisibility);
    }
}

File: lib/entities/overlay_notification.dart
class OverlayNotification {
  final String title;
  final String content;
  final NotificationVisibility notificationVisibility;

  const OverlayNotification(this.title, this.content, this.notificationVisibility);

  static const noNotification = OverlayNotification("", "", NotificationVisibility.none);

  // Method to serialize OverlayNotification object to JSON
  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'content': content,
      'notificationVisibility': notificationVisibility.name,
    };
  }

  // Factory method to deserialize JSON to OverlayNotification object
  factory OverlayNotification.fromJson(Map<String, dynamic> json) {
    return OverlayNotification(
      json['title'] as String,
      json['content'] as String,
      _getNotificationVisibility(json['notificationVisibility']),
    );
  }

  // Helper method to convert string to NotificationVisibility enum
  static NotificationVisibility _getNotificationVisibility(String value) {
    switch (value) {
      case 'visibilityPublic':
        return NotificationVisibility.visibilityPublic;
      case 'visibilitySecret':
        return NotificationVisibility.visibilitySecret;
      case 'visibilityPrivate':
        return NotificationVisibility.visibilityPrivate;
      case 'none':
        return NotificationVisibility.none;
      default:
        return NotificationVisibility.none; // Default to none if invalid value
    }
  }
}

enum NotificationVisibility {
  /// Show this notification in its entirety on all lockscreens.
  visibilityPublic,

  /// Do not reveal any part of this notification on a secure lockscreen.
  visibilitySecret,

  /// Show this notification on all lockscreens, but conceal sensitive or private information on secure lockscreens.
  visibilityPrivate,
  none,
}
